research/02-repl.tex

```diff
@@ -1,5 +1,172 @@
+open columns}
+\begin{table}[]
+\centering
+\begin{tabular}{lccccccccc}
+                                  & \rot{Python} & \rot{IPython} & \rot{R} & \rot{\shortstack[c]{Common\\Lisp}} & \rot{Haskell} & \rot{Swift} & \rot{AutoCAD} \\
+\toprule
+Executes single expressions       & \cmark       &               & \cmark  & \cmark                             & \cmark        & \cmark      &               \\
+Executes statements               & \cmark       &               & \cmark  & \cmark                             & \cmark        & \cmark      &               \\
+Input \& output history           & \cmark       &               & \cmark  & \cmark                             & \cmark        & \cmark      &               \\
+Persistent input history          & \cmark       &               & \cmark  & \xmark                             & \cmark        & \cmark      &               \\
+Multiline input editing           & \cmark       &               & \cmark  & \cmark                             & \cmark        & \cmark      &               \\
+Redefining identifiers            & \cmark       &               & \cmark  & N/A                                & \cmark        & \cmark      &               \\
+Error reporting                   & \cmark       &               & \cmark  & \cmark                             & \cmark        & \cmark      &               \\
+Context-sensitive code completion & \cmark       &               & \xmark  & N/A                                & \xmark        & \cmark      &               \\
+Help or documentation system      & \cmark       &               & \cmark  & \cmark                             & \xmark        & \xmark      &               \\
+Additional commands to the REPL   & \xmark       &               & \xmark  & \cmark                             & \cmark        & \cmark      &               \\
+Nested REPLs to enable debugging  & \xmark       &               & \xmark  & \cmark                             & \xmark        & \xmark      &               \\
+\bottomrule
+\end{tabular}
+\caption{A feature comparison of several well-known REPLs}
+\label{table:feature-matrix}
+\end{table}
+
+\paragraph{Input \& output history} REPLs keep a history of inputs and outputs,
+such that previously entered expressions can be retrieved together with their
+(evaluated) results. This can be with, say, a keyboard shortcut to retrieve
+prevsiouly entered inputs and with variables bound to previously yielded values.
+
+\paragraph{Persistent history} The input and output history as discussed
+previously can be recorded into a file (either per-project or globally) to
+enable a persistent history of input and output.
+
+\paragraph{Multiline input editing} Constructs in a programming language are
+not necessarily bound to one line; a method for example often has the following
+structure:
+\begin{lstlisting}
+<Type> <ID> ([<ID> ","]*) {
+	[<Statement>]+
+}
+\end{lstlisting}
```

Comment:
In this case I don't think being language agnostic is useful, as it's just an example.